S2D1 - CPS Error handling

piergm opened the pull request:
This PR introduces ResponseValidator, a utility class for index resolution validation in CPS. It:  

- Enforces consistent error handling for cross-project index operations based on IndicesOptions
- Handles both flat (unqualified) and qualified index expressions
- Provides appropriate error responses for missing/closed indices and authorization 
- Processes wildcard patterns that resolve differently across projects

piergm committed:
impl

piergm committed:
Update docs/changelog/135358.yaml

piergm committed:
Merge branch 'elastic:main' into S2D1-error-handling

piergm committed:
iter

piergm committed:
Merge branch 'elastic:main' into S2D1-error-handling

elasticsearchmachine commented:
Pinging @elastic/es-search-foundations (Team:Search Foundations)

piergm committed:
javadoc

piergm committed:
Merge branch 'main' into S2D1-error-handling

piergm committed:
better handing of qualified expressions

benchaplin reviewed (commented):
Thanks for the helpful javadocs! Left a couple small comments - still struggling to understand the meat of the logic so maybe I'll come back later...

piergm committed:
iter

piergm committed:
iter

piergm committed:
Merge branch 'main' into S2D1-error-handling

piergm committed:
Merge branch 'main' into S2D1-error-handling

piergm committed:
updated tests

piergm committed:
updated tests

piergm committed:
updated tests

piergm committed:
refactor CPS error util

piergm committed:
logic rework

piergm committed:
Merge branch 'main' into S2D1-error-handling

piergm committed:
deleted unused class

piergm committed:
t push
:Merge branch 'S2D1-error-handling' of github.com:piergm/elasticsearch into S2D1-error-handling

piergm committed:
renames + static utility

justzh reviewed (commented):
This looks fine.

piergm committed:
cleaner code + java docs

piergm committed:
cleaner

piergm committed:
Merge branch 'main' into S2D1-error-handling

piergm committed:
iter

piergm committed:
reworked

piergm committed:
iter

n1v0lg commented on server/src/main/java/org/elasticsearch/transport/ResponseValidator.java:
```diff
@@ -0,0 +1,217 @@
+        // Check resolution result
+        if (matchingExpression.localIndexResolutionResult() == SUCCESS) {
+            // Successfully found if there are concrete expressions
+            return matchingExpression.expressions().isEmpty() ? new IndexNotFoundException(remoteExpression) : null;
```
Same as for the local case, think we need `allow_no_indices` check here

n1v0lg commented on server/src/main/java/org/elasticsearch/transport/ResponseValidator.java:
```diff
@@ -0,0 +1,217 @@
+        // Find the matching expression in the linked project
+        ResolvedIndexExpression.LocalExpressions matchingExpression = findMatchingExpression(resolvedExpressionsInProject, resource);
+        if (matchingExpression == null) {
+            return new IndexNotFoundException(remoteExpression);
```
I _think_ this should never happen, right? Correct me if I'm wrong. Otherwise, lets add an assertion

piergm replied:
Correct, I just wanted to be extra sure instead of throwing NPE, now I see you added an assert false, LGTM!

n1v0lg commented on server/src/main/java/org/elasticsearch/transport/ResponseValidator.java:
```diff
@@ -0,0 +1,217 @@
+        ResolvedIndexExpression.LocalIndexResolutionResult result,
+        String expression
+    ) {
+        if (result == CONCRETE_RESOURCE_NOT_VISIBLE) {
```
We need to account for ignore_unavailable here

n1v0lg commented on server/src/main/java/org/elasticsearch/transport/ResponseValidator.java:
```diff
@@ -0,0 +1,217 @@
+                        return e;
+                    }
+                } else {
+                    // qualified linked project expression
```
We also need to check the local expression here still

n1v0lg reviewed (commented):
Nice -- I like the last version -- few more comments around handling IndicesOptions

piergm committed:
Delete docs/changelog/135358.yaml

piergm committed:
iter

piergm committed:
Merge branch 'S2D1-error-handling' of github.com:piergm/elasticsearch into S2D1-error-handling

piergm committed:
optimized lookup

piergm commented on server/src/main/java/org/elasticsearch/search/crossproject/ResponseValidator.java:
```diff
@@ -0,0 +1,217 @@
+        ResolvedIndexExpressions projectExpressions,
+        String resource
+    ) {
+        return projectExpressions.expressions()
```
Done here: https://github.com/elastic/elasticsearch/pull/135358/commits/8c66dd0e07651691e2237cb96f852a7f024d959d

piergm committed:
Revert "optimized lookup"

This reverts commit 8c66dd0e07651691e2237cb96f852a7f024d959d.

n1v0lg reviewed (approved):
LGTM ❤️ thanks for iterating on this! Before merging, lets also update the PR description as it's slightly out of date now.

piergm committed:
Merge branch 'main' into S2D1-error-handling
